### DISCUSSION TITLE
Building from Source to incorporate delegate.

### DISCUSSION THREAD

**Discussion by Budgie2** [ImageMagick/ImageMagick#3249] [Help]
Hi, I have run into a problem with trying to build from source.  It appears that the source comes with many delegates not set.  

I am not a coder as I have explained, so am well out of my depth here but building from scratch on a laptop and including the .fpx delegate (libfpx) which I also built from source, 

I was able to install with the delegate fpx installed but jpg was not installed.  Is there another source I should have used or have I missed switches in the configuration.
Grateful for help here please.

**Comment by fmw42** [comment]
What platform?  Why are you building from source and not just using a binary? Binaries comes with delegates. With source, you have to install delegates manually before installing ImageMagick.

**Reply by Budgie2** [reply]
I am building on openSUSE Tumbleweed.  I am building from source because I could not find the fpx delegate.
On last occasion when I built from source I didn't have the same problem so may have missed a step but of course I did build and include the fpx delegate first time.  I would still like to get this build to work.  Part of my learning experience!

**Comment by urban-warrior** [comment]
ImageMagick. for certain formats, requires external delegate libraries.  To validate these libraries, you must have both the object library but also any headers associated with the library (e.g. jpeg.h).   Take a look at the output of your `configure` command-line.  It will notify you which delegate libraries were validated.  If a particular library is not validated, check `config.log` to determine why it failed to validate.

**Reply by Budgie2** [reply]
Not at the machine right now but will look again this evening.  Many thanks,
Budge.

**Reply by Budgie2** [reply]
I confess I am stuck.  Looked at the config.log and am none the wiser as there is so much which is not working but I cannot see why jpeg is not installed.  The headers are all in the coders subdirectory which is included in the download.  I tried including the specific inclusion of jpeg, which is not required as far as I can see but whatever I do I cannot get jpeg to be included in delegates.
After various tries this is what I ran:-
`./configure --with-modules --with-fpx --with-jpeg` 

and this is what I received:-
`ImageMagick 7.0.11-0 is configured as follows. Please verify that this
configuration matches your expectations.

Host system type: x86_64-pc-linux-gnu
Build system type: x86_64-pc-linux-gnu

                  Option                        Value
------------------------------------------------------------------------------
Shared libraries  --enable-shared=yes           yes
Static libraries  --enable-static=yes           yes
Build utilities   --with-utilities=yes          yes
Module support    --with-modules=yes            yes
GNU ld            --with-gnu-ld=yes             yes
Quantum depth     --with-quantum-depth=16       16
High Dynamic Range Imagery
                  --enable-hdri=yes             yes

Install documentation:                          yes

Memory allocation library:
  JEMalloc          --with-jemalloc=no          no
  TCMalloc          --with-tcmalloc=no          no
  UMem              --with-umem=no              no

Delegate library configuration:
  BZLIB             --with-bzlib=yes            yes
  Autotrace         --with-autotrace=no         no
  DJVU              --with-djvu=yes             no
  DPS               --with-dps=yes              no
  FFTW              --with-fftw=yes             no
  FLIF              --with-flif=yes             no
  FlashPIX          --with-fpx=yes              yes
  FontConfig        --with-fontconfig=yes       no
  FreeType          --with-freetype=yes         no
  Ghostscript lib   --with-gslib=no             no
  Graphviz          --with-gvc=yes              no
  HEIC              --with-heic=yes             no
  JBIG              --with-jbig=yes             no
  JPEG v1           --with-jpeg=yes             no
  JPEG XL           --with-jxl=no               no
  LCMS              --with-lcms=yes             no
  LQR               --with-lqr=yes              no
  LTDL              --with-ltdl=yes             yes
  LZMA              --with-lzma=yes             no
  Magick++          --with-magick-plus-plus=yes yes
  OpenEXR           --with-openexr=yes          no
  OpenJP2           --with-openjp2=yes          no
  PANGO             --with-pango=yes            no
  PERL              --with-perl=no              no
  PNG               --with-png=yes              no
  RAQM              --with-raqm=yes             no
  RAW               --with-raw=yes              no
  RSVG              --with-rsvg=no              no
  TIFF              --with-tiff=yes             no
  WEBP              --with-webp=yes             no
  WMF               --with-wmf=yes              no
  X11               --with-x=                   no
  XML               --with-xml=yes              no
  ZIP               --with-zip=yes              no
  ZLIB              --with-zlib=yes             no
  ZSTD              --with-zstd=yes             no

Delegate program configuration:
  GhostPCL          None                        pcl6 (unknown)
  GhostXPS          None                        gxps (unknown)
  Ghostscript       None                        gs (9.53.3)

Font configuration:
  Apple fonts       --with-apple-font-dir=default
  Dejavu fonts      --with-dejavu-font-dir=default      none
  Ghostscript fonts --with-gs-font-dir=default          /usr/share/ghostscript/fonts/
  URW-base35 fonts  --with-urw-base35-font-dir=default  none
  Windows fonts     --with-windows-font-dir=default     none

X11 configuration:
  X_CFLAGS        = 
  X_PRE_LIBS      = 
  X_LIBS          = 
  X_EXTRA_LIBS    = 

Options used to compile and link:
  PREFIX          = /usr/local
  EXEC-PREFIX     = /usr/local
  VERSION         = 7.0.11-0
  CC              = gcc
  CFLAGS          = -fopenmp -Wall -g -O2 -mtune=ivybridge -fexceptions -pthread -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
  CPPFLAGS        =  -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
  PCFLAGS         = 
  DEFS            = -DHAVE_CONFIG_H
  LDFLAGS         = 
  LIBS            =               -lbz2    -lltdl  -lm    -lpthread 
  CXX             = g++
  CXXFLAGS        =  -pthread
  FEATURES        = DPC HDRI Cipher OpenMP Modules
  DELEGATES       = bzlib fpx ps video
==============================================================================

`
I did try using the binaries but had problems with opensuse and missing dependencies which frustrated me using rpm or yast so I need to get the build right and would appreciate further guidance when you have time please.

**Reply by Budgie2** [reply]
Sorry, that copy and paste went wrong.  Here is another try:-
`
                  Option                        Value
------------------------------------------------------------------------------
Shared libraries  --enable-shared=yes           yes
Static libraries  --enable-static=yes           yes
Build utilities   --with-utilities=yes          yes
Module support    --with-modules=yes            yes
GNU ld            --with-gnu-ld=yes             yes
Quantum depth     --with-quantum-depth=16       16
High Dynamic Range Imagery
                  --enable-hdri=yes             yes

Install documentation:                          yes

Memory allocation library:
  JEMalloc          --with-jemalloc=no          no
  TCMalloc          --with-tcmalloc=no          no
  UMem              --with-umem=no              no

Delegate library configuration:
  BZLIB             --with-bzlib=yes            yes
  Autotrace         --with-autotrace=no         no
  DJVU              --with-djvu=yes             no
  DPS               --with-dps=yes              no
  FFTW              --with-fftw=yes             no
  FLIF              --with-flif=yes             no
  FlashPIX          --with-fpx=yes              yes
  FontConfig        --with-fontconfig=yes       no
  FreeType          --with-freetype=yes         no
  Ghostscript lib   --with-gslib=no             no
  Graphviz          --with-gvc=yes              no
  HEIC              --with-heic=yes             no
  JBIG              --with-jbig=yes             no
  JPEG v1           --with-jpeg=yes             no
  JPEG XL           --with-jxl=no               no
  LCMS              --with-lcms=yes             no
  LQR               --with-lqr=yes              no
  LTDL              --with-ltdl=yes             yes
  LZMA              --with-lzma=yes             no
  Magick++          --with-magick-plus-plus=yes yes
  OpenEXR           --with-openexr=yes          no
  OpenJP2           --with-openjp2=yes          no
  PANGO             --with-pango=yes            no
  PERL              --with-perl=no              no
  PNG               --with-png=yes              no
  RAQM              --with-raqm=yes             no
  RAW               --with-raw=yes              no
  RSVG              --with-rsvg=no              no
  TIFF              --with-tiff=yes             no
  WEBP              --with-webp=yes             no
  WMF               --with-wmf=yes              no
  X11               --with-x=                   no
  XML               --with-xml=yes              no
  ZIP               --with-zip=yes              no
  ZLIB              --with-zlib=yes             no
  ZSTD              --with-zstd=yes             no

Delegate program configuration:
  GhostPCL          None                        pcl6 (unknown)
  GhostXPS          None                        gxps (unknown)
  Ghostscript       None                        gs (9.53.3)

Font configuration:
  Apple fonts       --with-apple-font-dir=default
  Dejavu fonts      --with-dejavu-font-dir=default      none
  Ghostscript fonts --with-gs-font-dir=default          /usr/share/ghostscript/fonts/
  URW-base35 fonts  --with-urw-base35-font-dir=default  none
  Windows fonts     --with-windows-font-dir=default     none

X11 configuration:
  X_CFLAGS        = 
  X_PRE_LIBS      = 
  X_LIBS          = 
  X_EXTRA_LIBS    = 

Options used to compile and link:
  PREFIX          = /usr/local
  EXEC-PREFIX     = /usr/local
  VERSION         = 7.0.11-0
  CC              = gcc
  CFLAGS          = -fopenmp -Wall -g -O2 -mtune=ivybridge -fexceptions -pthread -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
  CPPFLAGS        =  -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
  PCFLAGS         = 
  DEFS            = -DHAVE_CONFIG_H
  LDFLAGS         = 
  LIBS            =               -lbz2    -lltdl  -lm    -lpthread 
  CXX             = g++
  CXXFLAGS        =  -pthread
  FEATURES        = DPC HDRI Cipher OpenMP Modules
  DELEGATES       = bzlib fpx ps video
`

**Comment by snibgo** [comment]
Your config.log for IM should contain lines like this:
```
configure:31656: checking for JPEG
configure:31658: result: 
:
:
:
configure:31773: checking if JPEG package is complete
configure:31786: result: yes
```
I suppose yours doesn't. What does it say about JPEG?

**Reply by Budgie2** [reply]
Hi snibgo and many thanks for the reply.  I was warned about headers but thought I had these resolved.  Clearly not because when I looked for jpeg this is what I have:-
`alastair@localhost:~/Downloads/ImageMagick-7.0.11-0> grep jpeg config.log
configure:30477: checking jpeglib.h usability
conftest.c:154:10: fatal error: jpeglib.h: No such file or directory
  154 | #include <jpeglib.h>
| #include <jpeglib.h>
configure:30477: checking jpeglib.h presence
conftest.c:121:10: fatal error: jpeglib.h: No such file or directory
  121 | #include <jpeglib.h>
| #include <jpeglib.h>
configure:30477: checking for jpeglib.h
configure:30485: checking for jpeg_read_header in -ljpeg
configure:30510: gcc -o conftest -fopenmp -Wall -g -O2 -mtune=ivybridge -fexceptions -pthread -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16  -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16  conftest.c -ljpeg  -lltdl -lbz2 -lgomp -lm  -lpthread >&5
/usr/lib64/gcc/x86_64-suse-linux/10/../../../../x86_64-suse-linux/bin/ld: cannot find -ljpeg
| char jpeg_read_header ();
| return jpeg_read_header ();
  JPEG v1           --with-jpeg=yes             no
ac_cv_header_jpeglib_h=no
ac_cv_lib_jpeg_jpeg_read_header=no
alastair@localhost:~/Downloads/ImageMagick-7.0.11-0> `

where should I get jpeglib
What have I missed please?

**Comment by snibgo** [comment]
Sorry, I know nothing about how to install packages on openSUSE.

But that is the problem. IM is expecting that you have installed the jpeg library, including headers such as jpeglib.h. But IM can't find these.

**Comment by Budgie2** [comment]
Hi snibgo, I have started over but I am not too familiar with setting some of the variables.  I have tried doing all the work in a new ImageMagic subdirectory in my home directory which is of course in my PATH environment and AFAIK everything should work from there or the other subdirectories within the PATH environment such as /usr/local/bin and so on.
I still cannot get the jpeg components working.  If you put aside the openSUSE package and think Debian,  what should I run with ./configure to include jpeg?  I have many delegates showing yes as being configured but not enabled.  I must be missing a switch.
What do you do whn you build from source.  
I have looked at AppImage but my system is not working with any of my AppImages at the moment.  This is an openSUSE problem, I hope short term.  I shall look again at binary packages but although rpm packages they may baulk at openSUSE.  Will try the binaries  but building really should, by definition, be totally compatible.
Any light you can shine would be appreciated.  I shall also ask on my openSUSE forum just in case there is an IM enthusiast interested.
Thanks anyway.
Budge

**Comment by snibgo** [comment]
I show how I build IM at [Compiling IM with Cygwin](http://im.snibgo.com/compim.htm).

I don't know if you need automake and autoconf before ./configure.
